pkg/simple/client/devops/jenkins/internal/github.go

```diff
@@ -0,0 +1,142 @@
+	source.CreateAttr("plugin", "github-branch-source")
+	source.CreateElement("id").SetText(githubSource.ScmId)
+	source.CreateElement("credentialsId").SetText(githubSource.CredentialId)
+	source.CreateElement("repoOwner").SetText(githubSource.Owner)
+	source.CreateElement("repository").SetText(githubSource.Repo)
+	if githubSource.ApiUri != "" {
+		source.CreateElement("apiUri").SetText(githubSource.ApiUri)
+	}
+	traits := source.CreateElement("traits")
+	if githubSource.DiscoverBranches != 0 {
+		traits.CreateElement("org.jenkinsci.plugins.github__branch__source.BranchDiscoveryTrait").
+			CreateElement("strategyId").SetText(strconv.Itoa(githubSource.DiscoverBranches))
+	}
+	if githubSource.DiscoverPRFromOrigin != 0 {
+		traits.CreateElement("org.jenkinsci.plugins.github__branch__source.OriginPullRequestDiscoveryTrait").
+			CreateElement("strategyId").SetText(strconv.Itoa(githubSource.DiscoverPRFromOrigin))
+	}
+	if githubSource.DiscoverPRFromForks != nil {
+		forkTrait := traits.CreateElement("org.jenkinsci.plugins.github__branch__source.ForkPullRequestDiscoveryTrait")
+		forkTrait.CreateElement("strategyId").SetText(strconv.Itoa(githubSource.DiscoverPRFromForks.Strategy))
+		trustClass := "org.jenkinsci.plugins.github_branch_source.ForkPullRequestDiscoveryTrait$"
+		if prTrust := GitHubPRDiscoverTrust(githubSource.DiscoverPRFromForks.Trust); prTrust.IsValid() {
+			trustClass += prTrust.String()
+		} else {
+			klog.Warningf("invalid GitHub discover PR trust value: %d", prTrust.Value())
+		}
+		forkTrait.CreateElement("trust").CreateAttr("class", trustClass)
+	}
+	if githubSource.DiscoverTags {
+		traits.CreateElement("org.jenkinsci.plugins.github__branch__source.TagDiscoveryTrait")
+	}
+	if githubSource.CloneOption != nil {
+		cloneExtension := traits.CreateElement("jenkins.plugins.git.traits.CloneOptionTrait").CreateElement("extension")
+		cloneExtension.CreateAttr("class", "hudson.plugins.git.extensions.impl.CloneOption")
+		cloneExtension.CreateElement("shallow").SetText(strconv.FormatBool(githubSource.CloneOption.Shallow))
+		cloneExtension.CreateElement("noTags").SetText(strconv.FormatBool(false))
+		cloneExtension.CreateElement("honorRefspec").SetText(strconv.FormatBool(true))
+		cloneExtension.CreateElement("reference")
+		if githubSource.CloneOption.Timeout >= 0 {
+			cloneExtension.CreateElement("timeout").SetText(strconv.Itoa(githubSource.CloneOption.Timeout))
+		} else {
+			cloneExtension.CreateElement("timeout").SetText(strconv.Itoa(10))
+		}
+
+		if githubSource.CloneOption.Depth >= 0 {
+			cloneExtension.CreateElement("depth").SetText(strconv.Itoa(githubSource.CloneOption.Depth))
+		} else {
+			cloneExtension.CreateElement("depth").SetText(strconv.Itoa(1))
+		}
+	}
+	if githubSource.RegexFilter != "" {
+		regexTraits := traits.CreateElement("jenkins.scm.impl.trait.RegexSCMHeadFilterTrait")
+		regexTraits.CreateAttr("plugin", "scm-api@2.4.0")
```

Comment:
Why does it need a hard code version? If another version, will there be a problem?

Reply:
I just refractory these code lines. You can see it from the master branch. https://github.com/kubesphere/kubesphere/blob/91c2e05616ef7abcdb51bf774df02f8eed803030/pkg/simple/client/devops/jenkins/pipeline_internal.go#L291

I can do some tests to see it still works well after removing the specified version.

Reply:
It works well. See https://github.com/kubesphere/kubesphere/pull/3177/commits/e4782b37f0cd007952a8e466f3667bd4ba8a5496#diff-bc352fa9668f5879136b80d2ab0539b58c06a6988bdb8282c34878a845621d86R74